File: clients/kmp/shared/src/commonTest/kotlin/org/wyrdsekai/app/engine/between/WarmHandoffTest.kt
package org.wyrdsekai.app.engine.between

import kotlinx.coroutines.test.runTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

class WarmHandoffTest {

    @Test
    fun sendHandoffPublishesToBetween() = runTest {
        val between = InMemoryBetweenClient()
        between.connect("ws://test")
        val manager = WarmHandoffManager(between, "node-1", "family-1")

        val context = WarmHandoffContext(
            fromDid = "bud-1",
            toDid = "bud-2",
            activeRoomId = "nexus",
            openConversationDids = listOf("bud-2"),
            recentTurns = listOf(
                ConversationTurn("user", "Hello", 1000),
                ConversationTurn("assistant", "Hi there!", 1001),
            ),
            vitalitySnapshot = mapOf("energy" to 0.8f, "confidence" to 0.7f),
            currentTask = "exploring",
            timestamp = 2000,
        )

        manager.sendHandoff(context, "node-2")

        assertEquals(1, between.published.size)
        val (subject, _) = between.published[0]
        assertTrue(subject.contains("family-1"))
        assertTrue(subject.contains("node-1"))
        assertTrue(subject.contains("node-2"))
        assertTrue(subject.contains("soul.handoff"))
    }

    @Test
    fun receiveHandoffTriggersCallback() = runTest {
        val between = InMemoryBetweenClient()
        between.connect("ws://test")
        val manager = WarmHandoffManager(between, "node-2", "family-1")

        var received: WarmHandoffContext? = null
        manager.onHandoffReceived { received = it }
        manager.startListening()

        // Simulate handoff from node-1
        val context = WarmHandoffContext(
            fromDid = "bud-1",
            toDid = "bud-2",
            activeRoomId = "terminal",
            timestamp = 3000,
        )
        val data = kotlinx.serialization.json.Json.encodeToString(
            WarmHandoffContext.serializer(), context
        ).encodeToByteArray()
        between.publish("between.household.family-1.node-1.node-2.soul.handoff", data)

        assertNotNull(received)
        assertEquals("terminal", received!!.activeRoomId)
        assertEquals("bud-1", received!!.fromDid)

        manager.stopListening()
    }

    @Test
    fun contextSerializesWithConversationTurns() {
        val context = WarmHandoffContext(
            fromDid = "bud-1",
            toDid = "bud-2",
            activeRoomId = "nexus",
            recentTurns = listOf(
                ConversationTurn("user", "How are you?", 100),
                ConversationTurn("assistant", "I'm well, thank you!", 101),
            ),
            vitalitySnapshot = mapOf("energy" to 0.5f),
            timestamp = 200,
        )

        val json = kotlinx.serialization.json.Json.encodeToString(
            WarmHandoffContext.serializer(), context
        )
        val restored = kotlinx.serialization.json.Json.decodeFromString(
            WarmHandoffContext.serializer(), json
        )
        assertEquals(2, restored.recentTurns.size)
        assertEquals("How are you?", restored.recentTurns[0].content)
    }

    @Test
    fun handoffSubjectMatchesPattern() = runTest {
        val between = InMemoryBetweenClient()
        between.connect("ws://test")
        val manager = WarmHandoffManager(between, "node-1", "family-abc")

        val context = WarmHandoffContext(
            fromDid = "bud-1", toDid = "bud-2",
            activeRoomId = "nexus", timestamp = 100,
        )
        manager.sendHandoff(context, "node-2")

        val subject = between.published[0].first
        assertEquals("between.household.family-abc.node-1.node-2.soul.handoff", subject)
    }
}

File: clients/rn/__tests__/engine/warm-handoff.test.ts
import { InMemoryBetweenClient } from '../../src/engine/between/BetweenClient';
import { WarmHandoffManager, type WarmHandoffContext } from '../../src/engine/between/WarmHandoff';

describe('WarmHandoffManager', () => {
  it('sendHandoff publishes to Between', async () => {
    const between = new InMemoryBetweenClient();
    await between.connect('ws://test');
    const manager = new WarmHandoffManager(between, 'node-1', 'family-1');

    const context: WarmHandoffContext = {
      fromDid: 'bud-1',
      toDid: 'bud-2',
      activeRoomId: 'nexus',
      openConversationDids: ['bud-2'],
      recentTurns: [
        { role: 'user', content: 'Hello', timestamp: 1000 },
        { role: 'assistant', content: 'Hi there!', timestamp: 1001 },
      ],
      vitalitySnapshot: { energy: 0.8, confidence: 0.7 },
      currentTask: 'exploring',
      timestamp: 2000,
    };

    manager.sendHandoff(context, 'node-2');

    expect(between.published.length).toBe(1);
    const { subject } = between.published[0];
    expect(subject).toContain('family-1');
    expect(subject).toContain('node-1');
    expect(subject).toContain('node-2');
    expect(subject).toContain('soul.handoff');
  });

  it('receive handoff triggers callback', async () => {
    const between = new InMemoryBetweenClient();
    await between.connect('ws://test');
    const manager = new WarmHandoffManager(between, 'node-2', 'family-1');

    let received: WarmHandoffContext | null = null;
    manager.onHandoffReceived(ctx => { received = ctx; });
    manager.startListening();

    const context: WarmHandoffContext = {
      fromDid: 'bud-1',
      toDid: 'bud-2',
      activeRoomId: 'terminal',
      openConversationDids: [],
      recentTurns: [],
      vitalitySnapshot: {},
      currentTask: null,
      timestamp: 3000,
    };
    const data = new TextEncoder().encode(JSON.stringify(context));
    between.publish('between.household.family-1.node-1.node-2.soul.handoff', data);

    expect(received).not.toBeNull();
    expect(received!.activeRoomId).toBe('terminal');
    expect(received!.fromDid).toBe('bud-1');

    manager.stopListening();
  });

  it('handoff subject matches pattern', async () => {
    const between = new InMemoryBetweenClient();
    await between.connect('ws://test');
    const manager = new WarmHandoffManager(between, 'node-1', 'family-abc');

    const context: WarmHandoffContext = {
      fromDid: 'bud-1', toDid: 'bud-2',
      activeRoomId: 'nexus', openConversationDids: [],
      recentTurns: [], vitalitySnapshot: {},
      currentTask: null, timestamp: 100,
    };
    manager.sendHandoff(context, 'node-2');

    expect(between.published[0].subject).toBe(
      'between.household.family-abc.node-1.node-2.soul.handoff',
    );
  });
});
